File: samples/_shared/ios/LLMAdapter.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//
// ─────────────────────────────────────────────────────────────────────────
// THE ONE FILE YOU EDIT WHEN YOU WANT A DIFFERENT LLM RUNTIME.
//
// Every Dazzle sample — chat-memory, chat-iot, chat-kb — links against
// this file. The ChatView, the DazzleEdge.chatAgent wiring, and the
// Dazzle storage calls are all identical no matter which adapter you
// pick. That's the whole point of the LLMClient protocol.
//
// Five adapters ship in the SDK:
//
//     A. LlamaCppClient         — any GGUF model, on-device
//     B. LiteRtLmClient         — Google's .litertlm format, on-device
//                                 (we ship the iOS port; nobody else
//                                 has LiteRT-LM running on iPhone)
//     C. FoundationModelsClient — Apple Intelligence, iOS 26+ / macOS 26+
//     D. OpenAICompatibleClient — anything that speaks the OpenAI REST
//                                 shape (OpenAI, HuggingFace Router,
//                                 Ollama, vLLM, Groq, Together, …)
//     E. AnthropicClient        — Anthropic Claude API directly
//                                 (`/v1/messages`, distinct shape from
//                                 OpenAI — `system` field, tool use as
//                                 content blocks, etc.)
//
// Default: LlamaCppClient with Qwen 2.5 1.5B Instruct (Q4_K_M), ~1 GB.
// Download it once with `samples/_scripts/download_models.sh`.
// ─────────────────────────────────────────────────────────────────────────

import Foundation

/// Build the `LLMClient` every sample's ChatAgent drives. Uncomment the
/// adapter you want; comment out the rest.
///
/// Convenience auto-detection: if any of the cloud API-key env vars
/// is set at launch time, switch to the matching cloud adapter
/// without code edits — useful for the smoke runner. Otherwise we
/// fall through to the default (LlamaCpp).
func makeLLMClient() async throws -> any LLMClient {
    let env = ProcessInfo.processInfo.environment
    if let key = env["ANTHROPIC_API_KEY"], !key.isEmpty {
        // ─── E auto ─── Anthropic when ANTHROPIC_API_KEY is in env.
        return AnthropicClient(
            model:     env["ANTHROPIC_MODEL"] ?? "claude-haiku-4-5-20251001",
            apiKey:    key,
            maxTokens: 1024,
            temperature: 0.3)
    }
    if let key = env["OPENAI_API_KEY"], !key.isEmpty {
        return OpenAICompatibleClient(
            baseURL: URL(string: "https://api.openai.com/v1")!,
            model:   env["OPENAI_MODEL"] ?? "gpt-4o-mini",
            apiKey:  key,
            temperature: 0.3,
            maxTokens: 512)
    }
    if let key = env["HF_TOKEN"], !key.isEmpty {
        return OpenAICompatibleClient(
            baseURL: URL(string: "https://router.huggingface.co/v1")!,
            model:   env["HF_MODEL"] ?? "meta-llama/Llama-3.3-70B-Instruct",
            apiKey:  key,
            temperature: 0.3,
            maxTokens: 512)
    }

    // ─── A ─── llama.cpp — any GGUF model (Llama 3, Gemma, Qwen, Phi,
    //                       DeepSeek, Mistral, …).
    //          Default: Qwen2.5-1.5B-Instruct-Q4_K_M (fast on A14+).
    //          Download: samples/_scripts/download_models.sh
    return try await LlamaCppClient(
        modelURL: ModelSetup.qwenGgufURL,
        systemPrompt: "You are a helpful on-device AI assistant.",
        temperature: 0.3,
        maxTokens: 512,
        nThreads: 4)

    // ─── B ─── Google LiteRT-LM (.litertlm)
    //          Default: Gemma 4 E2B IT.
    // return try await LiteRtLmClient(
    //     modelURL: ModelSetup.gemmaLiteRtURL,
    //     systemPrompt: "You are a helpful on-device AI assistant.",
    //     temperature: 0.3,
    //     maxTokens: 512)

    // ─── C ─── Apple Intelligence (iOS 26+ / macOS 26+)
    //          Device-side, no downloads, no API keys. Requires
    //          FoundationModels eligibility (Apple silicon 15+, iPhone
    //          15 Pro and later).
    // if #available(iOS 26.0, macOS 26.0, *),
    //    FoundationModelsClient.isAvailable {
    //     return FoundationModelsClient(
    //         systemPrompt: "You are a helpful on-device AI assistant.",
    //         temperature: 0.3,
    //         maxTokens: 512)
    // }

    // ─── D ─── OpenAI-compatible endpoint.
    //
    // Supports: OpenAI proper, HuggingFace Inference Router, Ollama,
    // vLLM, Groq, Together, llama-server, and anything that speaks
    // `POST /v1/chat/completions` with the same JSON shape.
    //
    // Two common configurations — uncomment the one you need and set
    // the matching API key in your environment.
    //
    //   export OPENAI_API_KEY=sk-...
    //   or
    //   export HF_TOKEN=hf_...
    //
    // OpenAI:
    // let key = ProcessInfo.processInfo.environment["OPENAI_API_KEY"] ?? ""
    // return OpenAICompatibleClient(
    //     baseURL: URL(string: "https://api.openai.com/v1")!,
    //     model:   "gpt-4o-mini",
    //     apiKey:  key,
    //     temperature: 0.3,
    //     maxTokens: 512)
    //
    // HuggingFace Router (free tier supports many open-weights LLMs):
    // let key = ProcessInfo.processInfo.environment["HF_TOKEN"] ?? ""
    // return OpenAICompatibleClient(
    //     baseURL: URL(string: "https://router.huggingface.co/v1")!,
    //     model:   "meta-llama/Llama-3.3-70B-Instruct",
    //     apiKey:  key,
    //     temperature: 0.3,
    //     maxTokens: 512)

    // ─── E ─── Anthropic (Claude) — `/v1/messages` API.
    //
    //   Distinct from OpenAI shape: `system` is top-level, tool use
    //   lives as `content` blocks, schema goes under `input_schema`.
    //   The SDK handles the mapping for you.
    //
    // let key = ProcessInfo.processInfo.environment["ANTHROPIC_API_KEY"] ?? ""
    // return AnthropicClient(
    //     model:     "claude-3-5-sonnet-latest",
    //     apiKey:    key,
    //     maxTokens: 1024,
    //     temperature: 0.3)
}

/// Helper: where the downloaded models live inside the app bundle.
///
/// The `_scripts/download_models.sh` script drops the files in
/// `samples/_scripts/_models/` on your Mac. Each sample's Xcode build
/// phase copies them into the .app bundle under `Resources/` so the
/// device can mmap them directly.
enum ModelSetup {
    /// Qwen 2.5 1.5B Instruct, Q4_K_M quantisation. ~1.0 GB.
    static var qwenGgufURL: URL {
        guard let url = Bundle.main.url(
                forResource: "qwen2.5-1.5b-instruct-q4_k_m",
                withExtension: "gguf")
        else {
            fatalError("""
                Qwen GGUF not bundled. Run:
                    samples/_scripts/download_models.sh
                then rebuild the sample.
                """)
        }
        return url
    }

    /// Gemma 4 E2B IT, LiteRT format. ~2.4 GB.
    static var gemmaLiteRtURL: URL {
        guard let url = Bundle.main.url(
                forResource: "gemma4-e2b-it",
                withExtension: "litertlm")
        else {
            fatalError("""
                Gemma LiteRT model not bundled. Run:
                    samples/_scripts/download_models.sh
                then rebuild the sample.
                """)
        }
        return url
    }
}

File: samples/_shared/android/LLMAdapter.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//
// ─────────────────────────────────────────────────────────────────────────
// THE ONE FILE YOU EDIT WHEN YOU WANT A DIFFERENT LLM RUNTIME.
//
// Every Dazzle sample — chat-memory, chat-iot, chat-kb — links against
// this file. The ChatScreen, the DazzleEdge.chatAgent wiring, and the
// Dazzle storage calls are all identical no matter which adapter you
// pick. That's the whole point of the LLMClient interface.
//
// Five adapters ship in the SDK:
//
//     A. LlamaCppClient         — any GGUF model, on-device
//     B. LiteRtLmClient         — Google's .litertlm format, on-device
//                                 (Android + iOS — the iOS bridge is
//                                 our own port; nobody else ships
//                                 LiteRT-LM for iOS today)
//     C. FoundationModelsClient — iOS-only. Not applicable on Android.
//     D. OpenAICompatibleClient — anything that speaks the OpenAI REST
//                                 shape (OpenAI, HuggingFace Router,
//                                 Ollama, vLLM, Groq, Together, …)
//     E. AnthropicClient        — Anthropic Claude API directly
//                                 (`/v1/messages`, distinct shape
//                                 from OpenAI — `system` field, tool
//                                 use as content blocks, etc.)
//
// Default: LlamaCppClient with Qwen 2.5 1.5B Instruct (Q4_K_M), ~1 GB.
// Run `samples/_scripts/download_models.sh` once to fetch the weights
// and `push_models_to_device.sh` to push them to /data/local/tmp on the
// device. See each sample's README for the setup.
// ─────────────────────────────────────────────────────────────────────────

package dev.dazzle.samples.shared

import android.content.Context
import dev.dazzle.sdk.LLMClient
import dev.dazzle.sdk.edge.AnthropicClient
import dev.dazzle.sdk.edge.LiteRtLmClient
import dev.dazzle.sdk.edge.LlamaCppClient
import dev.dazzle.sdk.edge.OpenAICompatibleClient
import java.io.File

object LLMAdapter {

    /// Build the `LLMClient` every sample's ChatAgent drives. Uncomment
    /// the adapter you want; comment out the rest.
    fun makeLLMClient(context: Context): LLMClient {

        // ─── A ─── llama.cpp — any GGUF model (Llama 3, Gemma, Qwen,
        //                       Phi, DeepSeek, Mistral, …).
        //          Default: Qwen 2.5 0.5B Instruct Q4_K_M (~400 MB).
        //
        // The 1.5B variant (~1 GB) produces better prose but is too
        // slow on mid-tier devices (e.g. Moto G35 takes 2–3 min per
        // turn). The 0.5B fits the "fast on phone" bucket the paper
        // targets and still grounds answers correctly once Dazzle
        // hands it the retrieved rows. Swap the path below if you
        // want the 1.5B — `ModelSetup.qwen15GgufFile(context)`.
        return LlamaCppClient(
            modelFile = ModelSetup.qwenGgufFile(context),
            systemPrompt = "You are a helpful on-device AI assistant.",
            temperature = 0.3f,
            maxTokens = 192,
            nThreads = 4,
        )

        // ─── B ─── Google LiteRT-LM (.litertlm)
        //          Default: Gemma 4 E2B IT.
        // return LiteRtLmClient(
        //     modelFile = ModelSetup.gemmaLiteRtFile(context),
        //     context = context,
        //     systemPrompt = "You are a helpful on-device AI assistant.",
        //     temperature = 0.3,
        //     maxTokens = 512,
        // )

        // ─── D ─── OpenAI-compatible endpoint.
        //
        // Supports: OpenAI proper, HuggingFace Inference Router, Ollama,
        // vLLM, Groq, Together, llama-server, and anything that speaks
        // `POST /v1/chat/completions` with the same JSON shape.
        //
        // Reads the key from a system property so you don't have to
        // wire BuildConfig:  `-Pdazzle.openai_key=sk-...` on the gradle
        // command line, or set it from your IDE run config. Empty key
        // is fine for endpoints that don't authenticate (Ollama,
        // llama-server, local vLLM).
        //
        // OpenAI:
        // return OpenAICompatibleClient(
        //     baseURL  = "https://api.openai.com/v1",
        //     model    = "gpt-4o-mini",
        //     apiKey   = System.getProperty("dazzle.openai_key", ""),
        //     temperature = 0.3,
        //     maxTokens   = 512,
        // )
        //
        // HuggingFace Router (free tier supports many open-weights LLMs):
        // return OpenAICompatibleClient(
        //     baseURL  = "https://router.huggingface.co/v1",
        //     model    = "meta-llama/Llama-3.3-70B-Instruct",
        //     apiKey   = System.getProperty("dazzle.hf_token", ""),
        //     temperature = 0.3,
        //     maxTokens   = 512,
        // )

        // ─── E ─── Anthropic (Claude) — `/v1/messages` API.
        //
        //   Distinct from OpenAI shape: `system` is top-level, tool
        //   use lives as `content` blocks, schema goes under
        //   `input_schema`. The SDK handles the mapping for you.
        //
        // return AnthropicClient(
        //     model     = "claude-3-5-sonnet-latest",
        //     apiKey    = System.getProperty("dazzle.anthropic_key", ""),
        //     maxTokens = 1024,
        //     temperature = 0.3,
        // )
    }
}

/// Helper: where the downloaded model files live on device.
///
/// `samples/_scripts/download_models.sh` drops them on your Mac; the
/// push step copies them to `/data/local/tmp/` on the connected device
/// (adb pushes need an unlocked device + USB debugging). The path is
/// readable by the app sandbox because `/data/local/tmp/` is granted
/// read access via the manifest's requestLegacyExternalStorage.
/// For production apps, ship the model inside assets or download it
/// through `ModelDownloader`.
object ModelSetup {

    /// Default: Qwen 2.5 0.5B Instruct Q4_K_M — ~400 MB, fast on
    /// mid-tier devices. Pulled from
    ///   HuggingFace: Qwen/Qwen2.5-0.5B-Instruct-GGUF
    /// Path ordering: pushed-by-adb first, then per-app files dir.
    fun qwenGgufFile(context: Context): File {
        val candidates = listOf(
            File("/data/local/tmp/qwen2.5-0.5b-instruct-q4_k_m.gguf"),
            File(context.filesDir, "qwen2.5-0.5b-instruct-q4_k_m.gguf"),
            File(context.getExternalFilesDir(null),
                 "qwen2.5-0.5b-instruct-q4_k_m.gguf"),
        )
        return candidates.firstOrNull { it.exists() }
            ?: error(
                "Qwen GGUF not found. Push it with:\n" +
                "  adb push samples/_scripts/_models/" +
                "qwen2.5-0.5b-instruct-q4_k_m.gguf /data/local/tmp/\n" +
                "or place it in the app's files dir."
            )
    }

    /// Larger variant (1.5B, ~1 GB). Better prose but slow on mid-tier
    /// devices — use when the target is flagships with >6 GB RAM.
    fun qwen15GgufFile(context: Context): File {
        val candidates = listOf(
            File("/data/local/tmp/qwen2.5-1.5b-instruct-q4_k_m.gguf"),
            File(context.filesDir, "qwen2.5-1.5b-instruct-q4_k_m.gguf"),
            File(context.getExternalFilesDir(null),
                 "qwen2.5-1.5b-instruct-q4_k_m.gguf"),
        )
        return candidates.firstOrNull { it.exists() }
            ?: error(
                "Qwen 1.5B GGUF not found. Push it with:\n" +
                "  adb push samples/_scripts/_models/" +
                "qwen2.5-1.5b-instruct-q4_k_m.gguf /data/local/tmp/\n" +
                "or use the 0.5B variant (faster on mid-tier phones)."
            )
    }

    fun gemmaLiteRtFile(context: Context): File {
        val candidates = listOf(
            File("/data/local/tmp/gemma4-e2b-it.litertlm"),
            File(context.filesDir, "gemma4-e2b-it.litertlm"),
        )
        return candidates.firstOrNull { it.exists() }
            ?: error(
                "Gemma LiteRT model not found. Push it with:\n" +
                "  adb push samples/_scripts/_models/gemma4-e2b-it.litertlm" +
                " /data/local/tmp/"
            )
    }
}
